Jellyfin.Server.Implementations/MediaSegments/MediaSegmentsManager.cs

```diff
@@ -0,0 +1,213 @@
+    private readonly IDbContextFactory<JellyfinDbContext> _dbProvider;
+    private bool _disposed;
+
+    /// <summary>
+    /// Initializes a new instance of the <see cref="MediaSegmentsManager"/> class.
+    /// </summary>
+    /// <param name="libraryManager">The library manager.</param>
+    /// <param name="dbProvider">The database provider.</param>
+    public MediaSegmentsManager(
+        ILibraryManager libraryManager,
+        IDbContextFactory<JellyfinDbContext> dbProvider)
+    {
+        _libraryManager = libraryManager;
+        _libraryManager.ItemRemoved += LibraryManagerItemRemoved;
+        _dbProvider = dbProvider;
+    }
+
+    /// <inheritdoc/>
+    public async Task<IReadOnlyList<MediaSegment>> CreateMediaSegments(Guid itemId, IReadOnlyList<MediaSegment> segments)
+    {
+        var item = _libraryManager.GetItemById(itemId);
+        if (item is null)
+        {
+            throw new InvalidOperationException("Item not found");
+        }
+
+        var dbContext = await _dbProvider.CreateDbContextAsync().ConfigureAwait(false);
+        await using (dbContext.ConfigureAwait(false))
+        {
+            foreach (var segment in segments)
+            {
+                segment.ItemId = itemId;
+                ValidateSegment(segment);
+
+                var found = await dbContext.Segments.FirstOrDefaultAsync(s => s.ItemId.Equals(segment.ItemId) && s.StreamIndex == segment.StreamIndex && s.Type == segment.Type && s.TypeIndex == segment.TypeIndex)
+                    .ConfigureAwait(false);
+
+                AddOrUpdateSegment(dbContext, segment, found);
+            }
+
+            await dbContext.SaveChangesAsync().ConfigureAwait(false);
+        }
+
+        return segments;
+    }
+
+    /// <inheritdoc/>
+    public async Task<IReadOnlyList<MediaSegment>> GetAllMediaSegments(Guid itemId, int? streamIndex = null, int? typeIndex = null, MediaSegmentType? type = null)
+    {
+        var item = _libraryManager.GetItemById(itemId);
+        if (item is null)
+        {
+            throw new InvalidOperationException("Item not found");
+        }
+
+        List<MediaSegment> allSegments;
+
+        var dbContext = await _dbProvider.CreateDbContextAsync().ConfigureAwait(false);
+        await using (dbContext.ConfigureAwait(false))
+        {
+            IQueryable<MediaSegment> queryable = dbContext.Segments.Where(s => s.ItemId.Equals(itemId));
+
+            if (streamIndex is not null)
+            {
+                queryable = queryable.Where(s => s.StreamIndex == streamIndex.Value);
+            }
+
+            if (type is not null)
+            {
+                queryable = queryable.Where(s => s.Type == type.Value);
+            }
+
+            if (!typeIndex.Equals(null))
+            {
+                queryable = queryable.Where(s => s.TypeIndex == typeIndex.Value);
+            }
+
+            allSegments = await queryable.AsNoTracking().ToListAsync().ConfigureAwait(false);
```

Comment:
No need for this to be a separate variable, you can just return this here and remove `allSegments`